include/dali/core/mm/async_pool.h

```diff
@@ -0,0 +1,428 @@
+      }
+      return f;
+    }
+    auto it = std::partition_point(pending.begin(), pending.end(), [&](pending_free *f) {
+      return !f->ready();
+    });
+    if (it == pending.end()) {
+      assert(!free.free_list.tail || !free.free_list.tail->ready());
+      return nullptr;
+    }
+    f = *it;
+    assert(f->ready());
+    return f;
+  }
+
+  void free_pending(PerStreamFreeBlocks &free) {
+    auto *f = find_first_ready(free);
+    while (f) {
+      global_pool_.deallocate(f->addr, f->bytes, f->alignment);
+      f = remove_pending_free(free, f);
+    }
+  }
+
+  void deallocate_async_impl(PerStreamFreeBlocks &free, char *ptr, size_t bytes, size_t alignment,
+                             cudaStream_t stream) {
+    auto *pending = add_pending_free(free.free_list, ptr, bytes, alignment, stream);
+    try {
+      free.by_size.insert({bytes, pending});
+    } catch (...) {
+      remove_pending_free(free.free_list, pending);
+      throw;
+    }
+  }
+
+  /**
+   * @brief Restores original block parameters (address, size and alignment) of a block
+   *        that was used for allocating a smaller block, possibly with stricter alignment.
+   */
+  void restore_original_params(char *&p, size_t &bytes, size_t &alignment) {
+    auto it = padded_.find(p);
+    if (it != padded_.end()) {
+      if (it->second.front_padding + bytes > it->second.bytes) {
+        throw std::invalid_argument("The deallocated memory points to a block that's smaller than "
+          "the size being freed. Check the size of the memory region being freed.");
+      }
+      p -= it->second.front_padding;
+      bytes = it->second.bytes;
+      alignment = it->second.alignment;
+      padded_.erase(it);
+    }
+  }
+
+  auto *add_pending_free(PendingFreeList &free, char *base, size_t bytes, size_t alignment,
+                         cudaStream_t stream) {
+    pending_free *f = FreeDescAlloc::allocate(1);
+    f = new (f)pending_free();
+    f->addr = base;
+    f->bytes = bytes;
+    f->alignment = alignment;
+    f->prev = nullptr;
+    f->next = free.head;
+    if (f->next)
+      f->next->prev = f;
+    free.head = f;
+    if (!free.tail) free.tail = f;
+    f->event = event_pool_.Get();
+    cudaEventRecord(f->event, stream);
+    num_pending_frees_++;
+    return f;
+  }
+
+  pending_free *remove_pending_free(PerStreamFreeBlocks &free, pending_free *f,
+                                    bool remove_by_size = true) {
+    if (remove_by_size)
+      free.by_size.erase({ f->bytes, f });
+    return remove_pending_free(free.free_list, f);
+  }
+
+  pending_free *remove_pending_free(PendingFreeList &free, pending_free *f) {
+    event_pool_.Put(std::move(f->event));
+    auto *prev = f->prev;
+    auto *next = f->next;
+    if (free.head == f)
+      free.head = next;
+    if (free.tail == f)
+      free.tail = prev;
+    if (prev) prev->next = next;
+    if (next) next->prev = prev;
+    *f = {};
+#if 0  // check list consistency
```

Comment:
I think we can enable it. This is for debug builds anyway, right?

Reply:
No, it's waaay too slow even for debug.
It was useful when I moved stuff around and had some problems with the list, but I do hope noone will ever need to re-enable it - but it's there so that when we see something fail we can reenable it quickly.

Reply:
How about `#ifdef DEBUG_LIST_CONSISTENCY` and add on top of the file `//#define DEBUG_LIST_CONSISTENCY` with an appropriate comment?

Reply:
Well, perhaps... if there are more things to fix.